src/Microsoft.Identity.Client/Instance/AuthorityEndpointResolutionManager.cs

```diff
@@ -58,32 +58,52 @@ public AuthorityEndpointResolutionManager(IServiceBundle serviceBundle, bool sho
             {
                 throw new MsalClientException(
                     MsalError.TenantDiscoveryFailedError,
-                    "Authorize endpoint was not found in the openid configuration");
+                    MsalErrorMessage.AuthorizeEndpointWasNotFoundInTheOpenIdConfiguration);
             }
 
             if (string.IsNullOrEmpty(edr.TokenEndpoint))
             {
                 throw new MsalClientException(
                     MsalError.TenantDiscoveryFailedError,
-                    "Token endpoint was not found in the openid configuration");
+                    MsalErrorMessage.TokenEndpointWasNotFoundInTheOpenIdConfiguration);
             }
 
             if (string.IsNullOrEmpty(edr.Issuer))
             {
                 throw new MsalClientException(
                     MsalError.TenantDiscoveryFailedError,
-                    "Issuer was not found in the openid configuration");
+                    MsalErrorMessage.IssuerWasNotFoundInTheOpenIdConfiguration);
             }
 
             endpoints = new AuthorityEndpoints(
-                edr.AuthorizationEndpoint.Replace("{tenant}", tenant),
-                edr.TokenEndpoint.Replace("{tenant}", tenant),
-                edr.Issuer.Replace("{tenant}", tenant));
+                edr.AuthorizationEndpoint.Replace(Constants.Tenant, tenant),
+                edr.TokenEndpoint.Replace(Constants.Tenant, tenant),
+                CreateSelfSignedJwtAudienceEndpointFromIssuer(edr, tenant));
 
             Add(authorityInfo, userPrincipalName, endpoints);
             return endpoints;
         }
 
+        internal string CreateSelfSignedJwtAudienceEndpointFromIssuer(TenantDiscoveryResponse endpoints, string tenant)

```

Comment:
This method needs explanation in my mind. It needs to say why this is needed

Reply:
i agree now that i look at it. how about `ReplaceNonTenantSpecificValueWithCorrectTenant`? 

---
In reply to: [298006468](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1237#discussion_r298006468) [](ancestors = 298006468)

Reply:
Like that much better. Perhaps just ReplaceNonTenantSpecificValueWithTenant

---
In reply to: [298222183](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1237#discussion_r298222183) [](ancestors = 298222183,298006468)